### PR TITLE
Add parameters attribute to schedules

### PR DESCRIPTION
**Pull request by lcapelluto**
### Summary
The `Schedule` class now has a `parameter` property and an `is_parameterized()` method. `Instruction`s and `Channel`s also get these attributes. This follows suit with `QuantumCircuit`, and also allows us to improve the efficiency of parameter assignment, which has been done for `Instructions.assign_parameters`.

### CODE REVIEW

**Review by taalexander — COMMENTED** [review]
Would it be possible to have `parameters` be returned as a list? This is important for enabling  `*args` based binding to a parameter for the pulse interface, eg, 

```
with pulse.build() as sched:
    pulse.call(param_sched, 0, 1, 2)
```

The other way around this would be to define `pulse.Function` which is a `signature` and a `Schedule` containing the parameters defined in `signature`.

**Comment by lcapelluto** [comment]
What you suggest is incompatible with the circuit interface: a parameter instance must be bound to a value. Order doesn't factor in

**Comment by taalexander** [comment]
> What you suggest is incompatible with the circuit interface: a parameter instance must be bound to a value. Order doesn't factor in. Please take a look

Thanks, I've taken a look, and have implemented a solution that at its core corresponds to a `Function(schedule, list(parameters))`, and `Call(function, arguments)`. This enables the reuse of a parameterized code block.

**Comment by lcapelluto** [comment]
for `Gaussian(..., amp=alpha*beta)`, which is the 0th parameter?

File: qiskit/pulse/channels.py [hunk 1]
```diff
@@ -59,6 +59,57 @@ def index(self) -> int:
...
+
+    def is_parameterized(self) -> bool:
+        """Return True iff the channel is parameterized."""
+        return bool(self.parameters)
+
+    def assign(self, parameter: Parameter, value: ParameterValueType) -> 'Channel':
+        """Return a new channel with the input Parameter assigned to value.
+
+        Args:
+            parameter (Parameter): A parameter in this expression whose value will be updated.
```
Comment (taalexander) [inline qiskit/pulse/channels.py:93]:
```suggestion
            parameter: A parameter in this expression whose value will be updated.
```

File: qiskit/pulse/channels.py [hunk 2]
```diff
@@ -59,6 +59,57 @@ def index(self) -> int:
...
+        if parameter not in self.parameters:
+            raise PulseError('Cannot bind parameters ({}) not present in the channel.'
+                             ''.format(parameter))
+
+        new_index = self.index.assign(parameter, value)
+        if not new_index.parameters:
+            self._validate_index(new_index)
+            new_index = int(new_index)
+
+        return type(self)(new_index)
```
Comment (taalexander) [inline qiskit/pulse/channels.py:111]: This returns a new channel, does this work with the other schedule parameter implementation? IIRC these mutated the schedule?

File: qiskit/pulse/instructions/play.py [hunk 3]
```diff
@@ -50,6 +50,12 @@ def __init__(self, pulse: Pulse,
         if name is None:
             name = pulse.name
         super().__init__((pulse, channel), pulse.duration, (channel,), name=name)
+        if pulse.is_parameterized():
```
Comment (taalexander) [inline qiskit/pulse/instructions/play.py:53]:
```suggestion
        
        if pulse.is_parameterized():
```

File: qiskit/pulse/library/pulse.py [hunk 4]
```diff
@@ -39,6 +39,10 @@ def id(self) -> int:  # pylint: disable=invalid-name
         """Unique identifier for this pulse."""
         return id(self)
 
+    def is_parameterized(self) -> bool:
```
Comment (taalexander) [inline qiskit/pulse/library/pulse.py:42]: This and the method below should be decorated with `abstractmethod`

File: qiskit/pulse/schedule.py [hunk 5]
```diff
@@ -61,9 +62,10 @@ def __init__(self, *schedules: Union[ScheduleComponent, Tuple[int, ScheduleCompo
         self._name = name
         self._duration = 0
 
+        # These attributes are populated by ``_mutable_insert``
         self._timeslots = {}
         self.__children = []
-
+        self._parameter_table = defaultdict(list)
```
Comment (taalexander) [inline qiskit/pulse/schedule.py:68]: This pattern is often repeated, is there a `ParameterTable` class?

File: qiskit/pulse/schedule.py [hunk 6]
```diff
@@ -650,23 +665,33 @@ def assign_parameters(self,
...
-                    if param in value_dict:
-                        new_index = new_channel.index.assign(param, value_dict[param])
-                        if not new_index.parameters:
-                            new_index = float(new_index)
-                            if float(new_index).is_integer():
-                                new_index = int(new_index)
-                        new_channel = type(chan)(new_index)
+                for param, value in value_dict.items():
+                    if param in new_channel.parameters:
+                        new_channel = new_channel.assign(param, value)
```
Comment (taalexander) [inline qiskit/pulse/schedule.py:694]: What happens if the old channel was used in an instruction? Doesn't this produce a new channel that isn't used anywhere except in timeslots?

**Review by taalexander — CHANGES_REQUESTED** [review]

File: qiskit/pulse/schedule.py (hunk 5 above)
Comment (lcapelluto) [reply qiskit/pulse/schedule.py:68]: Yes but unfortunately it doesn't work here. Maybe a community member could generalize it in the future.

File: qiskit/pulse/schedule.py (hunk 6 above)
Comment (lcapelluto) [reply qiskit/pulse/schedule.py:694]: I don't know what you are getting at. `DriveChannel(0)` is equal to `DriveChannel(0)`; it doesn't matter if they are the same instance

File: qiskit/pulse/channels.py (hunk 2 above)
Comment (lcapelluto) [reply qiskit/pulse/channels.py:111]: What other implementation?

File: qiskit/pulse/library/pulse.py (hunk 4 above)
Comment (SooluThomas) [reply qiskit/pulse/library/pulse.py:43]: Fixed in https://github.com/Qiskit/qiskit-terra/pull/5250/commits/9239d66ed6a9f9c3ded453045292092748e3c232

**Review by taalexander — APPROVED** [review]

**Pull request merged**